### PR TITLE
Sdks 330

### PR DESCRIPTION
**Pull request by witrisna**
SDKS-330 Centralize Login implementation
- Upgrade dependency libraries
- Centralize Login Implementation
- Centralize Login automation
- Only one App can register redirect URI

### CODE REVIEW

File: forgerock-auth/src/main/java/org/forgerock/android/auth/AppAuthConfigurer.java [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+import lombok.AccessLevel;
+import lombok.Getter;
+import lombok.RequiredArgsConstructor;
+
+/**
+ * AppAuth Integration, https://github.com/openid/AppAuth-Android
+ * this class provides AppAuth integration and customization
+ */
+@RequiredArgsConstructor
+public class AppAuthConfigurer {
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/AppAuthConfigurer.java:31]: Can this be package private?

File: forgerock-auth/src/main/java/org/forgerock/android/auth/AppAuthConfigurer.java [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            return new AuthorizationServiceConfiguration(
+                    Uri.parse(oAuth2Client.getAuthorizeUrl().toString()),
+                    Uri.parse(oAuth2Client.getTokenUrl().toString()));
+        } catch (MalformedURLException e) {
+            throw new RuntimeException(e);
+        }
+    };
+
+    /**
+     * Override the default OAuth2 endpoint defined under the String.xml
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/AppAuthConfigurer.java:57]: If this class is not designed to be package private, please complete the Javadoc for this method. This comment applies to other methods in this class.

File: forgerock-auth/src/main/java/org/forgerock/android/auth/AppAuthConfigurer.java [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    }
+
+    /**
+     * Finish up the AppAuth customization.
+     */
+    public FRUser.Browser done() {
+        return parent;
+    }
+
+
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/AppAuthConfigurer.java:99]: Extra line

File: forgerock-auth/src/main/java/org/forgerock/android/auth/AppAuthConfigurer.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+    /**
+     * Override the {@link AppAuthConfiguration} that prepared by the SDK.
+     */
+    public AppAuthConfigurer appAuthConfiguration(@NonNull Consumer<AppAuthConfiguration.Builder> authorizationService) {
+        this.appAuthConfigurationBuilder = authorizationService;
+        return this;
+    }
+
+    /**
+     * Override the {@link CustomTabsIntent} that prepared by the SDK.
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/AppAuthConfigurer.java:85]: 'that was prepared by the SDK.'?

File: forgerock-auth/src/main/java/org/forgerock/android/auth/AppAuthConfigurer.java [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+     * Override the default OAuth2 endpoint defined under the String.xml
+     * forgerock_url
+     */
+    public AppAuthConfigurer authorizationServiceConfiguration(@NonNull Supplier<AuthorizationServiceConfiguration> authorizationServiceConfiguration) {
+        this.authorizationServiceConfigurationSupplier = authorizationServiceConfiguration;
+        return this;
+    }
+
+    /**
+     * Override the {@link AuthorizationRequest} that prepared by the SDK.
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/AppAuthConfigurer.java:66]: Do you mean 'that was prepared by the SDK.'?

File: forgerock-auth/src/main/java/org/forgerock/android/auth/FRUser.java [hunk 6]
```diff
@@ -8,12 +8,29 @@
 package org.forgerock.android.auth;
 
 import android.content.Context;
+import android.content.Intent;
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/FRUser.java:11]: Update year on the class header

File: forgerock-auth/src/main/java/org/forgerock/android/auth/InterceptorHandler.java [hunk 7]
```diff
@@ -8,12 +8,13 @@
 package org.forgerock.android.auth;
 
 import android.content.Context;
-import lombok.AllArgsConstructor;
-import lombok.Builder;
-import lombok.Getter;
 
 import java.util.List;
 
+import lombok.Builder;
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/InterceptorHandler.java:14]: Update year on the class header

File: forgerock-auth/src/main/java/org/forgerock/android/auth/OAuth2Client.java [hunk 8]
```diff
@@ -202,6 +216,35 @@ public void onResponse(@NotNull Call call, @NotNull Response response) {
         }
     }
 
+    public void endSession(@NonNull String idToken, FRListener<Void> listener) {
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/OAuth2Client.java:219]: Missing Javadoc. Same comment may apply for others methods in this class.

File: forgerock-auth/src/main/java/org/forgerock/android/auth/RetrieveAccessTokenInterceptor.java [hunk 9]
```diff
@@ -21,22 +21,26 @@
...
-                        Listener.onSuccess(chain.getListener(), result);
-                    }
-
-                    @Override
-                    public void onException(Exception e) {
-                        //If cannot get the AccessToken, process to next interceptor in the chain.
-                        chain.proceed(sessionToken);
-                    }
-                });
+        tokenManager.getAccessToken(accessToken -> {
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/RetrieveAccessTokenInterceptor.java:24]: Update year on the class header

File: forgerock-auth/src/main/java/org/forgerock/android/auth/RetrieveSSOTokenInterceptor.java [hunk 10]
```diff
@@ -20,13 +19,8 @@
...
     @Override
     public void intercept(final Chain chain, final Void input) {
-
         SSOToken ssoToken = singleSignOnManager.getToken();
-        if (ssoToken != null) {
-            chain.proceed(ssoToken);
-        } else {
-            chain.getListener().onException(new AuthenticationRequiredException("Authentication Required"));
-        }
+        chain.proceed(ssoToken);
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/RetrieveSSOTokenInterceptor.java:23]: Update year on the class header

File: forgerock-auth/src/main/java/org/forgerock/android/auth/TokenManager.java [hunk 11]
```diff
@@ -25,16 +27,29 @@
      * Sends an authorization request to the authorization service.
      *
      * @param token    The SSO Token received with the result of {@link AuthService}
+     *
      * @param listener Listener that listens to changes resulting from OAuth endpoints .
      */
-    void exchangeToken(SSOToken token, final FRListener<AccessToken> listener);
+    void exchangeToken(SSOToken token, final Map<String, String> additionalParameters, final FRListener<AccessToken> listener);
```
Comment (rodrigoareis) [inline forgerock-auth/src/main/java/org/forgerock/android/auth/TokenManager.java:33]: Missing parameter on Javadoc

File: samples/auth/src/main/AndroidManifest.xml [hunk 12]
```diff
@@ -1,56 +1,74 @@
-<?xml version="1.0" encoding="utf-8"?>
-<!--
-  ~ Copyright (c) 2019 - 2020 ForgeRock. All rights reserved.
+<?xml version="1.0" encoding="utf-8"?><!--
+  ~ Copyright (c) 2019 ForgeRock. All rights reserved.
```
Comment (rodrigoareis) [inline samples/auth/src/main/AndroidManifest.xml:2]: Update year?

File: samples/auth/src/main/res/menu/menu.xml [hunk 13]
```diff
@@ -37,4 +37,5 @@
         android:icon="@android:drawable/ic_menu_call"
         android:title="@string/invoke"
         app:showAsAction="collapseActionView"/>
+
```
Comment (rodrigoareis) [inline samples/auth/src/main/res/menu/menu.xml:40]: Update year?

File: samples/auth/src/main/res/values/strings.xml [hunk 14]
```diff
@@ -10,6 +10,7 @@
     <string name="logout">Logout</string>
     <string name="instanceId">Show InstanceId</string>
     <string name="userinfo">UserInfo</string>
+    <string name="browser">Launch Browser</string>
```
Comment (rodrigoareis) [inline samples/auth/src/main/res/values/strings.xml:13]: Update year?

**Review by rodrigoareis — COMMENTED** [review]
Overall implementation looks good! Please, review the comments before the approval.

File: forgerock-auth/src/main/java/org/forgerock/android/auth/AppAuthConfigurer.java (hunk 1 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/AppAuthConfigurer.java:31]: This is a public interface for developers to override the appAuth default configuration.

File: forgerock-auth/src/main/java/org/forgerock/android/auth/AppAuthConfigurer.java (hunk 2 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/AppAuthConfigurer.java:57]: Updated

File: forgerock-auth/src/main/java/org/forgerock/android/auth/AppAuthConfigurer.java (hunk 3 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/AppAuthConfigurer.java:99]: Updated

File: forgerock-auth/src/main/java/org/forgerock/android/auth/AppAuthConfigurer.java (hunk 5 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/AppAuthConfigurer.java:66]: updated

File: forgerock-auth/src/main/java/org/forgerock/android/auth/FRUser.java (hunk 6 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/FRUser.java:11]: Updated

File: forgerock-auth/src/main/java/org/forgerock/android/auth/InterceptorHandler.java (hunk 7 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/InterceptorHandler.java:14]: Updated

File: forgerock-auth/src/main/java/org/forgerock/android/auth/OAuth2Client.java (hunk 8 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/OAuth2Client.java:219]: Updated.

File: forgerock-auth/src/main/java/org/forgerock/android/auth/RetrieveAccessTokenInterceptor.java (hunk 9 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/RetrieveAccessTokenInterceptor.java:24]: Updated

File: forgerock-auth/src/main/java/org/forgerock/android/auth/RetrieveSSOTokenInterceptor.java (hunk 10 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/RetrieveSSOTokenInterceptor.java:23]: Updated

File: forgerock-auth/src/main/java/org/forgerock/android/auth/TokenManager.java (hunk 11 above)
Comment (witrisna) [reply forgerock-auth/src/main/java/org/forgerock/android/auth/TokenManager.java:33]: Updated

File: samples/auth/src/main/AndroidManifest.xml (hunk 12 above)
Comment (witrisna) [reply samples/auth/src/main/AndroidManifest.xml:2]: Updated

File: samples/auth/src/main/res/menu/menu.xml (hunk 13 above)
Comment (witrisna) [reply samples/auth/src/main/res/menu/menu.xml:40]: Updated

File: samples/auth/src/main/res/values/strings.xml (hunk 14 above)
Comment (witrisna) [reply samples/auth/src/main/res/values/strings.xml:13]: Updated

**Review by rodrigoareis — APPROVED** [review]
Changes looks good to me

**Review by spetrov — APPROVED** [review]
Looks good to me!

**Review by spetrov — APPROVED** [review]

**Pull request merged**